internal/fileutil/watcher.go

```diff
@@ -31,50 +30,64 @@ func NewWatcher() *Watcher {
 	}
 }
 
-// Add adds a new watch.
-func (watcher *Watcher) Add(filePath string) {
+// Watch updates the watched file paths.
+func (watcher *Watcher) Watch(ctx context.Context, filePaths []string) {
 	watcher.mu.Lock()
 	defer watcher.mu.Unlock()
 
-	// already watching
-	if _, ok := watcher.watching[filePath]; ok {
-		return
-	}
-
-	ctx := log.WithContext(context.Background(), func(c zerolog.Context) zerolog.Context {
-		return c.Str("watch_file", filePath)
-	})
 	watcher.initLocked(ctx)
 
-	if watcher.eventWatcher != nil {
-		if err := watcher.eventWatcher.Add(filePath); err != nil {
-			log.Error(ctx).Msg("fileutil/watcher: failed to watch file with event-based file watcher")
+	var add []string
+	seen := map[string]struct{}{}
+	for _, filePath := range filePaths {
+		if _, ok := watcher.watching[filePath]; !ok {
```

Comment:
Do we know if there are any corner cases where we might still need to add a watch even if the same path already exists in the `watching` set?

If I add a file path to be watched (e.g. `foo`) and then do something like `rm foo && mv bar foo` or `mv foo bar && mv bar foo` will we lose the file watch? (Do we need to handle operations like that?)

Reply:
I'm not sure how the underlying library we use handles this. I will look at the source code.

Reply:
> A path will remain watched if it gets renamed to somewhere else on the same filesystem, but the monitor will get removed if the path gets deleted and re-created, or if it's moved to a different filesystem.

So the fsnotify watcher will miss these events.

It appears that he polling watcher will handle this though.

Reply:
I assume we can't just call Remove() and Add() for each path in `filePaths`, as that would leave a brief window where we might miss an event in between?

Would it make sense to add some logic to the "handle events" goroutine below to automatically remove a path from the `watching` set whenever it observes an event where `evt.Op.Has(fsnotify.Remove)`?

Reply:
Should we only using the polling watcher then?

Reply:
Ok I added a test for this and only use the polling watcher. It seems to detect events on newly created files.

Reply:
Sorry, I totally misunderstood that it always uses both the event-based watcher and the polling watcher (I thought it was either/or). I think either way should be fine.